### PR TITLE
implement enforced validation for form

### PR DESCRIPTION
**Pull request by Glignos**
Signed-off-by: Glignos <glignos93@gmail.com>

### CODE REVIEW

File: src/app/add-action/add-action.component.ts [hunk 1]
```diff
@@ -56,6 +58,14 @@ export class AddActionComponent {
   }
 
   openEditor() {
-    this.isEditorVisible = true;
+    if (!this.action.mainKey) {
+      this.toastr.error('Choose a key for the field to be added');
+      return;
```
Comment (harunurhan) [inline src/app/add-action/add-action.component.ts]: unnecessary

File: src/app/add-action/add-action.component.ts [hunk 2]
```diff
@@ -56,6 +58,14 @@ export class AddActionComponent {
...
+    } else if (!this.subSchema) {
+      this.toastr.error('Choose a key using autocompletion values for the field to be added');
+      return;
```
Comment (harunurhan) [inline src/app/add-action/add-action.component.ts]: ditto

File: src/app/app.module.ts [hunk 3]
```diff
@@ -75,7 +79,10 @@ export class RavenErrorHandler implements ErrorHandler {
     AlertModule.forRoot(),
     PaginationModule.forRoot(),
     JsonEditorModule,
-    TypeaheadModule.forRoot()
+    TypeaheadModule.forRoot(),
+    CommonModule,
+    BrowserAnimationsModule,
+    ToastrModule.forRoot()
```
Comment (harunurhan) [inline src/app/app.module.ts]: here's where you configure the module. Like https://github.com/inspirehep/record-editor/blob/master/src/app/app.module.ts#L51

File: src/app/multi-editor/multi-editor.component.ts [hunk 4]
```diff
@@ -175,10 +180,9 @@ export class MultiEditorComponent implements OnInit {
       })
       .catch(error => {
         if (error.json().message) {
-          this.totalRecords = -1;
-          this.errorMessage = error.json().message;
+          this.toastr.error(error.json().message);
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: no `.json()` again. see above comment.

File: src/app/multi-editor/multi-editor.component.ts [hunk 5]
```diff
@@ -175,10 +180,9 @@ export class MultiEditorComponent implements OnInit {
       })
       .catch(error => {
         if (error.json().message) {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `Response.json()` is actually parsing response body each time is called
https://github.com/angular/angular/blob/5.2.0/packages/http/src/body.ts#L26 you can see here.

So what you should do is
```typescript
const errorMessage = error.json().message;
if (errorMessage) {
    this.toastr.error(errorMessage);
}
```

File: src/app/multi-editor/multi-editor.component.ts [hunk 6]
```diff
@@ -219,11 +222,11 @@ export class MultiEditorComponent implements OnInit {
       },
       (error) => {
         if (error.json().message) {
-          this.totalRecords = -1;
-          this.errorMessage = error.json().message;
+          this.toastr.error(error.json().message);
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

Comment (harunurhan) [reply src/app/multi-editor/multi-editor.component.ts]: also it seems like showing error message can be a function and used here and above.

File: src/app/multi-editor/multi-editor.component.ts [hunk 7]
```diff
@@ -137,9 +136,17 @@ export class MultiEditorComponent implements OnInit {
   }
 
   previewActions() {
+    if (!this.schemaKeysStoreService.validateActionsKeypaths(this.userActions)) {
+      this.toastr.error('Please use valid paths provided by autocompletion');
+      this.changeDetectorRef.markForCheck();
+      return;
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

File: src/app/multi-editor/multi-editor.component.ts [hunk 8]
```diff
@@ -137,9 +136,17 @@ export class MultiEditorComponent implements OnInit {
...
+    } else if (!this.checkExistenceOfAction(this.userActions)) {
+      this.toastr.error('Please use at least one action to preview');
+      this.changeDetectorRef.markForCheck();
+      return;
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

File: src/app/multi-editor/multi-editor.component.ts [hunk 9]
```diff
@@ -264,6 +265,17 @@ export class MultiEditorComponent implements OnInit {
     this.changeDetectorRef.markForCheck();
   }
 
+  checkExistenceOfAction(userActions: UserActions): boolean {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: private.

File: src/app/multi-editor/multi-editor.component.ts [hunk 10]
```diff
@@ -264,6 +265,17 @@ export class MultiEditorComponent implements OnInit {
...
+    let existingAction = false;
+    userActions.actions.forEach(action => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: use `some` for this, it's like old style `for` with `break;` when the check condition is true for the first time.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

File: src/app/multi-editor/multi-editor.component.ts [hunk 11]
```diff
@@ -264,6 +265,17 @@ export class MultiEditorComponent implements OnInit {
...
+      if (action.mainKey) {
+        existingAction = true;
+      } else {
+        this.toastr.error('Please fill in the field for every action');
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: this will show toast for every empty action, and showing toastr also done after calling this based on the result.

Maybe a leftover? if it's necessary should be called after this function is called based the result.

File: src/app/shared/services/schema-keys-store.service.ts [hunk 12]
```diff
@@ -86,10 +88,14 @@ export class SchemaKeysStoreService {
       if (subSchema['type'] === 'object') {
         if (subSchema['properties'][splitPath[index]]) {
           subSchema = subSchema['properties'][splitPath[index]];
+        } else {
+          return null;
```
Comment (harunurhan) [inline src/app/shared/services/schema-keys-store.service.ts:92]: why is that necessary?

File: src/app/shared/services/schema-keys-store.service.ts [hunk 13]
```diff
@@ -109,5 +115,24 @@ export class SchemaKeysStoreService {
...
+      if (!this.findSubschema(action.mainKey)) {
+        invalidActions = true;
+      }
+    });
+    userActions.conditions.forEach(condition => {
+      if (!this.findSubschema(condition.key)) {
+        invalidActions = true;
+      }
+    });
+    if (invalidActions) {
```
Comment (harunurhan) [inline src/app/shared/services/schema-keys-store.service.ts]: `return !invalidActions`

File: src/app/shared/services/schema-keys-store.service.ts [hunk 14]
```diff
@@ -109,5 +115,24 @@ export class SchemaKeysStoreService {
     return subSchema;
   }
 
+  validateActionsKeypaths(userActions: UserActions): boolean {
+    let invalidActions = false;
+    userActions.actions.forEach(action => {
```
Comment (harunurhan) [inline src/app/shared/services/schema-keys-store.service.ts]: `some` ditto

File: src/app/shared/services/schema-keys-store.service.ts [hunk 15]
```diff
@@ -109,5 +115,24 @@ export class SchemaKeysStoreService {
...
+      if (!this.findSubschema(action.mainKey)) {
+        invalidActions = true;
+      }
+    });
+    userActions.conditions.forEach(condition => {
```
Comment (harunurhan) [inline src/app/shared/services/schema-keys-store.service.ts]: `some` ditto

Comment (harunurhan) [reply src/app/shared/services/schema-keys-store.service.ts]: also run only `if (!invalidActions)`

File: src/app/shared/services/schema-keys-store.service.ts [hunk 16]
```diff
@@ -109,5 +115,24 @@ export class SchemaKeysStoreService {
     return subSchema;
   }
 
+  validateActionsKeypaths(userActions: UserActions): boolean {
+    let invalidActions = false;
```
Comment (harunurhan) [inline src/app/shared/services/schema-keys-store.service.ts]: `invalid` is good enough.

File: src/app/shared/services/schema-keys-store.service.ts [hunk 17]
```diff
@@ -109,5 +115,24 @@ export class SchemaKeysStoreService {
     return subSchema;
   }
 
+  validateActionsKeypaths(userActions: UserActions): boolean {
```
Comment (harunurhan) [inline src/app/shared/services/schema-keys-store.service.ts]: `validateUserActions`

Comment (harunurhan) [reply src/app/shared/services/schema-keys-store.service.ts]: does this make sense in this service? fits better in the component or in a new service if it is likely to be reused in the near future.

File: src/app/multi-editor/multi-editor.component.ts [hunk 18]
```diff
@@ -137,9 +136,17 @@ export class MultiEditorComponent implements OnInit {
   }
 
   previewActions() {
+    if (!this.schemaKeysStoreService.validateActionsKeypaths(this.userActions)) {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: Could be better to just disable preview button instead maybe if it's invalid?

Also seems to me that you could make use of `@angular/forms` https://angular.io/guide/form-validation, better not to reinvent the wheel in a worse way, please check if forms could be used for your case, I think so though.

File: src/app/autocomplete-input/autocomplete-input.component.html [hunk 19]
```diff
@@ -1,15 +1,12 @@
 <div *ngIf="tagsEnabled then tagTemplate else defaultTemplate"></div>
-<div *ngIf="typeaheadNoResults === true">
-  <i class="fa fa-remove"></i> No Results Found
-</div>
 
 <ng-template #defaultTemplate>
-  <input [ngModel]="value" (ngModelChange)="modelChange($event)" [typeahead]="dataSource" (typeaheadNoResults)="typeaheadNoResults = $event"
-  (typeaheadOnSelect)="selectUserInput($event)" [disabled]="disabled" [attr.class]="className" [typeaheadMinLength]="0">
+  <input class="form-control input-sm" [ngModel]="value" (ngModelChange)="modelChange($event)" [typeahead]="dataSource" (typeaheadNoResults)="typeaheadNoResults = $event"
+  (typeaheadOnSelect)="selectUserInput($event)" [disabled]="disabled" [class.alert-danger]="typeaheadNoResults" [typeaheadMinLength]="0">
```
Comment (harunurhan) [inline src/app/autocomplete-input/autocomplete-input.component.html]: you can use `.danger` here too, instead of alert-danger

**Review by harunurhan — CHANGES_REQUESTED** [review]

File: src/app/shared/services/schema-keys-store.service.ts (hunk 12 above)
Comment (Glignos) [reply src/app/shared/services/schema-keys-store.service.ts:92]: If a key is invalid im returning a null value allowing me to know that the path was faulty

File: src/app/shared/services/schema-keys-store.service.ts (hunk 17 above)
Comment (Glignos) [reply src/app/shared/services/schema-keys-store.service.ts]: I dont think it's gonna be reused so im moving it to the component where its used

File: src/app/multi-editor/multi-editor.component.ts [hunk 20]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
     this.changeDetectorRef.markForCheck();
   }
 
+  private checkExistenceOfAction(userActions: UserActions): boolean {
+    let existingAction = false;
+    return userActions.actions.some((element, index, array) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: would be nice to have a new line with `.some`

Comment (harunurhan) [reply src/app/multi-editor/multi-editor.component.ts]: only first param is enough here, so with a renaming this would be `.some(action => {..`

File: src/app/multi-editor/multi-editor.component.ts [hunk 21]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
...
+      }
+    });
+  }
+
+  private showError(error: any) {
+    const errorMessage = error.json().message;
+    if (errorMessage) {
+      this.toastr.error(errorMessage);
+    } else {
+      this.toastr.error(error);
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: Is this a thing?

File: src/app/multi-editor/multi-editor.component.ts [hunk 22]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
...
+  private checkExistenceOfAction(userActions: UserActions): boolean {
+    let existingAction = false;
+    return userActions.actions.some((element, index, array) => {
+      if (element.mainKey) {
+        return true;
+      }
+    });
+  }
+
+  private showError(error: any) {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `error: Response`

Comment (harunurhan) [reply src/app/multi-editor/multi-editor.component.ts]: `displayErrorMessage`

File: src/app/multi-editor/multi-editor.component.ts [hunk 23]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
...
+    if (errorMessage) {
+      this.toastr.error(errorMessage);
+    } else {
+      this.toastr.error(error);
+    }
+  }
+
+  private validateActionsKeypaths(userActions: UserActions): boolean {
+    let invalid = false;
+    userActions.actions.some((element, index, array) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

File: src/app/multi-editor/multi-editor.component.ts [hunk 24]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
...
+      if (!this.schemaKeysStoreService.findSubschema(element.mainKey)) {
+        return true;
+      }
+    });
+    if (!invalid) {
+      userActions.conditions.forEach((element, index, array) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `some`

File: src/app/multi-editor/multi-editor.component.ts [hunk 25]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
...
+        if (!this.schemaKeysStoreService.findSubschema(element.key)) {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `.some(condition => !this.schemaKeysStoreService.findSubschema(element.key))` is better. Because returning boolean in an `if` block is always unnecessary since you can always return the `if` condition itself

File: src/app/multi-editor/multi-editor.component.ts (hunk 23 above)
Comment (harunurhan) [reply src/app/multi-editor/multi-editor.component.ts]: you should assign `some` result to `invalid`, this code shouldn't even work.

File: src/app/multi-editor/multi-editor.component.ts [hunk 26]
```diff
@@ -264,6 +251,40 @@ export class MultiEditorComponent implements OnInit {
...
+      this.toastr.error(errorMessage);
+    } else {
+      this.toastr.error(error);
+    }
+  }
+
+  private validateActionsKeypaths(userActions: UserActions): boolean {
+    let invalid = false;
+    userActions.actions.some((element, index, array) => {
+      if (!this.schemaKeysStoreService.findSubschema(element.mainKey)) {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto (check comment for `condition`)

**Review by harunurhan — CHANGES_REQUESTED** [review]

File: src/app/conditions/conditions.component.html [hunk 27]
```diff
@@ -12,7 +11,8 @@
       </select>
     </div>
     <div class="form-group">
-      <input *ngIf="!(condition.matchType == 'does not exist')"type="text" name="value" [(ngModel)]="condition.value" class="form-control input-sm" placeholder="condition value">
+      <input *ngIf="!(condition.matchType == 'does not exist')" type="text" name="value" [(ngModel)]="condition.value" class="form-control input-sm"
```
Comment (harunurhan) [inline src/app/conditions/conditions.component.html]: `===`

File: src/app/multi-editor/multi-editor.component.ts [hunk 28]
```diff
@@ -137,25 +135,27 @@ export class MultiEditorComponent implements OnInit {
...
-          this.totalRecords = -1;
-          this.errorMessage = error.json().message;
-        } else {
-          this.errorMessage = error;
-        }
-        this.changeDetectorRef.markForCheck();
-      });
+    if (!this.checkExistenceOfAction(this.userActions)) {
+      this.toastr.error('Please use at least one action to preview');
+      this.changeDetectorRef.markForCheck();
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ?

File: src/app/multi-editor/multi-editor.component.ts [hunk 29]
```diff
@@ -137,25 +135,27 @@ export class MultiEditorComponent implements OnInit {
...
+    } else if (!this.validateActionsKeypaths(this.userActions)) {
+      this.toastr.error('Please use valid paths provided by autocompletion');
+      this.changeDetectorRef.markForCheck();
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ?

File: src/app/multi-editor/multi-editor.component.ts [hunk 30]
```diff
@@ -137,25 +135,27 @@ export class MultiEditorComponent implements OnInit {
...
+          this.records = res.json_records;
+          this.jsonPatches = res.json_patches;
+          this.validationErrors = res.errors;
+          this.previewMode = true;
+          this.filterRecords(this.filterExpressions);
+          this.changeDetectorRef.markForCheck();
+        }),
+        ((error) => {
+          this.displayErrorMessage(error);
+          this.changeDetectorRef.markForCheck();
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ?

File: src/app/multi-editor/multi-editor.component.ts [hunk 31]
```diff
@@ -264,6 +250,28 @@ export class MultiEditorComponent implements OnInit {
...
+  }
+
+  private displayErrorMessage(error: Response) {
+    const errorMessage = error.json().message || 'Something went wrong :)';
+    this.toastr.error(errorMessage);
+  }
+
+  private validateActionsKeypaths(userActions: UserActions): boolean {
+    let invalid = false;
+    invalid = userActions.actions.some(action => !this.schemaKeysStoreService.findSubschema(action.mainKey));
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]:
```
let invalid = userActions.actions
    .some(...
```

File: src/app/multi-editor/multi-editor.component.ts [hunk 32]
```diff
@@ -264,6 +250,28 @@ export class MultiEditorComponent implements OnInit {
...
+    if (!invalid) {
+      invalid = userActions.conditions.some(action => !this.schemaKeysStoreService.findSubschema(action.key));
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `.some` to next line would be better

File: src/app/shared/services/query.service.ts [hunk 33]
```diff
@@ -36,25 +36,23 @@ export class QueryService {
...
 
-  save(userActions: UserActions, checkedRecords: string[], allSelected: boolean): Promise<SubmitMessage> {
+  save(userActions: UserActions, checkedRecords: string[], allSelected: boolean): Observable<SubmitMessage> {
     return this.http
       .post(`${this.url}/update`, {
         userActions,
         ids: checkedRecords,
         allSelected
       }
       ).map(res => res.json())
```
Comment (harunurhan) [inline src/app/shared/services/query.service.ts]: move to above like with `;`, it'll fail on linter.

File: src/app/multi-editor/multi-editor.component.ts [hunk 34]
```diff
@@ -165,21 +165,15 @@ export class MultiEditorComponent implements OnInit {
 
   private fetchPage() {
     this.queryService.fetchPaginatedRecords(this.currentPage, this.pageSize)
-      .toPromise()
-      .then((json) => {
+      .subscribe((json) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: no `(` `)` is needed actually for single params.

**Review by harunurhan — CHANGES_REQUESTED** [review]

File: src/app/multi-editor/multi-editor.component.ts [hunk 35]
```diff
@@ -264,6 +243,24 @@ export class MultiEditorComponent implements OnInit {
     this.changeDetectorRef.markForCheck();
   }
 
-}
+  private checkExistenceOfAction(userActions: UserActions): boolean {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `chekIfanyActionHasMainKey`
or
`hasAnyNonEmptyAction`
